doc/core/man/tap2rpm.1
.TH TAP2RPM "1" "July 2001" "" ""
.SH NAME
tap2rpm \- create RPM packages which wrap .tap files
.SH SYNOPSIS
.B tap2rpm
[options]
.SH DESCRIPTION
Create a set of RPM/SRPM packages in the current directory
.TP
\fB\-u\fR, \fB\--unsigned\fR 
do not sign the RPM package
.TP
\fB\-t\fR, \fB\--tapfile\fR \fI<tapfile>\fR
Build the application around the given .tap (default twistd.tap)
.TP
\fB\-y\fR, \fB\--type\fR \fI<type>\fR
The configuration has the given type . Allowable types are
\fBtap\fR, \fBsource\fR, \fBxml\fR and \fBpython\fR.
The first three types are \fBmktap(1)\fR output formats,
while the last one is a manual building of application 
(see \fBtwistd(1)\fR, the \fB\-y\fR option).
.TP
\fB\-p\fR, \fB\--protocol\fR \fI<protocol>\fR
The name of the protocol this will be used to serve. This is intended
as a part of the description. Default is the name of the tapfile, minus
any extensions.
.TP
\fB\-d\fR, \fB\--rpmfile\fR \fI<rpmfile>\fR
The name of the RPM package. Default is 'twisted-'+protocol.
.TP
\fB\-V\fR, \fB\--set-version\fR \fI<version>\fR
The version of the RPM package. The default is 1.0
.TP
\fB\-e\fR, \fB\--description\fR \fI<description>\fR
The one-line description. Default is uninteresting.
.TP
\fB\-l\fR, \fB\--long_description\fR \fI<long_description>\fR
A multi-line description. Default is explanation about
this being an automatic package created from tap2rpm.
.TP
\fB\-m\fR, \fB\--maintainer\fR \fI<maintainer>\fR
The maintainer, as "Name Lastname <email address>". This will
go in the meta-files, as well as be used as the id to sign the package.
.TP
\fB\--version\fR
Output version information and exit.
.SH AUTHOR
tap2rpm was written by Sean Reifschneider based on tap2deb by Moshe Zadka.
This man page is heavily based on the tap2deb man page by Moshe Zadka.
.SH "REPORTING BUGS"
To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
.SH COPYRIGHT
Copyright \(co 2000-2008 Twisted Matrix Laboratories.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
mktap(1)


NEWS
Ticket numbers in this file can be looked up by visiting
http://twistedmatrix.com/trac/ticket/<number>

Twisted Core 10.0.0 (2010-03-01)
================================

Features
--------
 - The twistd man page now has a SIGNALS section. (#689)

 - reactor.spawnProcess now will not emit a PotentialZombieWarning
   when called before reactor.run, and there will be no potential for
   zombie processes in this case. (#2078)

 - High-throughput applications based on Perspective Broker should now
   run noticably faster thanks to the use of a more efficient decoding
   function in Twisted Spread. (#2310)

 - Documentation for trac-post-commit-hook functionality in svn-dev
   policy. (#3867)

 - twisted.protocols.socks.SOCKSv4 now supports the SOCKSv4a protocol.
   (#3886)

 - Trial can now output test results according to the subunit
   protocol, as long as Subunit is installed (see
   https://launchpad.net/subunit). (#4004)

 - twisted.protocols.amp now provides a ListOf argument type which can
   be composed with some other argument types to create a zero or more
   element sequence of that type. (#4116)

 - If returnValue is invoked outside of a function decorated with
   @inlineCallbacks, but causes a function thusly decorated to exit, a
   DeprecationWarning will be emitted explaining this potentially
   confusing behavior.  In a future release, this will cause an
   exception. (#4157)

 - twisted.python.logfile.BaseLogFile now has a reopen method allowing
   you to use an external logrotate mechanism. (#4255)

Bugfixes
--------
 - FTP.ftp_NLST now handles requests on invalid paths in a way
   consistent with RFC 959. (#1342)

 - twisted.python.util.initgroups now calls the low-level C initgroups
   by default if available: the python version can create lots of I/O
   with certain authentication setup to retrieve all the necessary
   information. (#3226)

 - startLogging now does nothing on subsequent invocations, thus
   fixing a terrible infinite recursion bug that's only on edge case.
   (#3289)

 - Stringify non-string data to NetstringReceiver.sendString before
   calculating the length so that the calculated length is equal to
   the actual length of the transported data. (#3299)

 - twisted.python.win32.cmdLineQuote now correctly quotes empty
   strings arguments (#3876)

 - Change the behavior of the Gtk2Reactor to register only one source
   watch for each file descriptor, instead of one for reading and one
   for writing. In particular, it fixes a bug with Glib under Windows
   where we failed to notify when a client is connected. (#3925)

 - Twisted Trial no longer crashes if it can't remove an old
   _trial_temp directory.  (#4020)

 - The optional _c_urlarg extension now handles unquote("") correctly
   on platforms where malloc(0) returns NULL, such as AIX.  It also
   compiles with less warnings. (#4142)

 - On POSIX, child processes created with reactor.spawnProcess will no
   longer automatically ignore the signals which the parent process
   has set to be ignored. (#4199)

 - All SOCKSv4a tests now use a dummy reactor with a deterministic
   resolve method. (#4275)

 - Prevent extraneous server, date and content-type headers in proxy
   responses. (#4277)

Deprecations and Removals
-------------------------
 - twisted.internet.error.PotentialZombieWarning is now deprecated.
   (#2078)

 - twisted.test.time_helpers is now deprecated. (#3719)

 - The deprecated connectUDP method of IReactorUDP has now been
   removed. (#4075)

 - twisted.trial.unittest.TestCase now ignores the previously
   deprecated setUpClass and tearDownClass methods. (#4175)

Other
-----
 - #917, #2406, #2481, #2608, #2689, #2884, #3056, #3082, #3199,
   #3480, #3592, #3718, #3935, #4066, #4083, #4154, #4166, #4169,
   #4176, #4183, #4186, #4188, #4189, #4194, #4201, #4204, #4209,
   #4222, #4234, #4235, #4238, #4240, #4245, #4251, #4264, #4268,
   #4269, #4282


Twisted Conch 10.0.0 (2010-03-01)
=================================

Bugfixes
--------
 - twisted.conch.checkers.SSHPublicKeyDatabase now looks in the
   correct user directory for authorized_keys files. (#3984)
 - twisted.conch.ssh.SSHUserAuthClient now honors preferredOrder when
   authenticating. (#4266)

Other
-----
 - #2391, #4203, #4265


Twisted Lore 10.0.0 (2010-03-01)
================================

Other
-----
 - #4241


Twisted Mail 10.0.0 (2010-03-01)
================================

Bugfixes
--------
 - twisted.mail.smtp.ESMTPClient and
   twisted.mail.smtp.LOGINAuthenticator now implement the (obsolete)
   LOGIN SASL mechanism according to the draft specification. (#4031)

 - twisted.mail.imap4.IMAP4Client will no longer misparse all html-
   formatted message bodies received in response to a fetch command.
   (#4049)

 - The regression in IMAP4 search handling of "OR" and "NOT" terms has
   been fixed. (#4178)

Other
-----
 - #4028, #4170, #4200


Twisted Names 10.0.0 (2010-03-01)
=================================

Bugfixes
--------
 - twisted.names.root.Resolver no longer leaks UDP sockets while
   resolving names. (#970)

Deprecations and Removals
-------------------------
 - Several top-level functions in twisted.names.root are now
   deprecated. (#970)

Other
-----
 - #4066


Twisted Pair 10.0.0 (2010-03-01)
================================

Other
-----
 - #4170


Twisted Runner 10.0.0 (2010-03-01)
==================================

Other
-----
 - #3961


Twisted Web 10.0.0 (2010-03-01)
===============================

Features
--------
 - Twisted Web in 60 Seconds, a series of short tutorials with self-
   contained examples on a range of common web topics, is now a part
   of the Twisted Web howto documentation. (#4192)

Bugfixes
--------
 - Data and File from twisted.web.static and
   twisted.web.distrib.UserDirectory will now only generate a 200
   response for GET or HEAD requests.
   twisted.web.client.HTTPPageGetter will no longer ignore the case of

twisted/scripts/tap2rpm.py
# -*- test-case-name: twisted.scripts.test.test_tap2rpm -*-
# Copyright (c) 2003-2010 Twisted Matrix Laboratories.
# See LICENSE for details.

import sys, os, shutil, time, glob
import subprocess
import tempfile
import tarfile
from StringIO import StringIO

from twisted.python import usage, log


#################################
#  data that goes in /etc/inittab
initFileData = '''\
#!/bin/sh
#
#  Startup script for a Twisted service.
#
#  chkconfig: - 85 15
#  description: Start-up script for the Twisted service "%(tap_file)s".

PATH=/usr/bin:/bin:/usr/sbin:/sbin

pidfile=/var/run/%(rpm_file)s.pid
rundir=/var/lib/twisted-taps/%(rpm_file)s/
file=/etc/twisted-taps/%(tap_file)s
logfile=/var/log/%(rpm_file)s.log

#  load init function library
. /etc/init.d/functions

[ -r /etc/default/%(rpm_file)s ] && . /etc/default/%(rpm_file)s

#  check for required files
if [ ! -x /usr/bin/twistd ]
then
	echo "$0: Aborting, no /usr/bin/twistd found"
	exit 0
fi
if [ ! -r "$file" ]
then
	echo "$0: Aborting, no file $file found."
	exit 0
fi

#  set up run directory if necessary
if [ ! -d "${rundir}" ]
then
	mkdir -p "${rundir}"
fi


case "$1" in
	start)
		echo -n "Starting %(rpm_file)s: twistd"
		daemon twistd  \\
				--pidfile=$pidfile \\
				--rundir=$rundir \\
				--%(twistd_option)s=$file \\
				--logfile=$logfile
		status %(rpm_file)s
		;;

	stop)
		echo -n "Stopping %(rpm_file)s: twistd"
		kill `cat "${pidfile}"`
		status %(rpm_file)s
		;;

	restart)
		"${0}" stop
		"${0}" start
		;;

    *)
		echo "Usage: ${0} {start|stop|restart|}" >&2
		exit 1
		;;
esac

exit 0
'''

#######################################
#  the data for creating the spec file
specFileData = '''\
Summary:    %(description)s
Name:       %(rpm_file)s
Version:    %(version)s
Release:    1
License:    Unknown
Group:      Networking/Daemons
Source:     %(tarfile_basename)s
BuildRoot:  %%{_tmppath}/%%{name}-%%{version}-root
Requires:   /usr/bin/twistd
BuildArch:  noarch

%%description
%(long_description)s

%%prep
%%setup
%%build

%%install
[ ! -z "$RPM_BUILD_ROOT" -a "$RPM_BUILD_ROOT" != '/' ] \
		&& rm -rf "$RPM_BUILD_ROOT"
mkdir -p "$RPM_BUILD_ROOT"/etc/twisted-taps
mkdir -p "$RPM_BUILD_ROOT"/etc/init.d
mkdir -p "$RPM_BUILD_ROOT"/var/lib/twisted-taps
cp "%(tap_file)s" "$RPM_BUILD_ROOT"/etc/twisted-taps/
cp "%(rpm_file)s.init" "$RPM_BUILD_ROOT"/etc/init.d/"%(rpm_file)s"

%%clean
[ ! -z "$RPM_BUILD_ROOT" -a "$RPM_BUILD_ROOT" != '/' ] \
		&& rm -rf "$RPM_BUILD_ROOT"

%%post
/sbin/chkconfig --add %(rpm_file)s
/sbin/chkconfig --level 35 %(rpm_file)s
/etc/init.d/%(rpm_file)s start

%%preun
/etc/init.d/%(rpm_file)s stop
/sbin/chkconfig --del %(rpm_file)s

%%files
%%defattr(-,root,root)
%%attr(0755,root,root) /etc/init.d/%(rpm_file)s
%%attr(0660,root,root) /etc/twisted-taps/%(tap_file)s

%%changelog
* %(date)s %(maintainer)s
- Created by tap2rpm: %(rpm_file)s (%(version)s)
'''

###############################
class MyOptions(usage.Options):
    optFlags = [["unsigned", "u"], ['quiet', 'q']]
    optParameters = [
                     ["tapfile", "t", "twistd.tap"],
                     ["maintainer", "m", "tap2rpm"],
                     ["protocol", "p", None],
                     ["description", "e", None],
                     ["long_description", "l",
                         "Automatically created by tap2rpm"],
                     ["set-version", "V", "1.0"],
                     ["rpmfile", "r", None],
                     ["type", "y", "tap", "type of configuration: 'tap', 'xml, "
                      "'source' or 'python'"],
                    ]

    #zsh_altArgDescr = {"foo":"use this description for foo instead"}
    #zsh_multiUse = ["foo", "bar"]
    #zsh_mutuallyExclusive = [("foo", "bar"), ("bar", "baz")]
    zsh_actions = {"type":"(tap xml source python)",
                   "rpmfile":'_files -g "*.rpm"'}
    #zsh_actionDescr = {"logfile":"log file name", "random":"random seed"}

    def postOptions(self):
        """
        Calculate the default values for certain command-line options.
        """
        # Options whose defaults depend on other parameters.
        if self['protocol'] is None:
            base_tapfile = os.path.basename(self['tapfile'])
            self['protocol'] = os.path.splitext(base_tapfile)[0]
        if self['description'] is None:
            self['description'] = "A TCP server for %s" % (self['protocol'],)
        if self['rpmfile'] is None:
            self['rpmfile'] = 'twisted-%s' % (self['protocol'],)

        # Values that aren't options, but are calculated from options and are
        # handy to have around.
        self['twistd_option'] = type_dict[self['type']]
        self['release-name'] = '%s-%s' % (self['rpmfile'], self['set-version'])



type_dict = {
    'tap': 'file',
    'python': 'python',
    'source': 'source',
    'xml': 'xml',
}



##########################
def makeBuildDir():
    """
    Set up the temporary directory for building RPMs.

    Returns: buildDir, a randomly-named subdirectory of baseDir.
    """
    tmpDir = tempfile.mkdtemp()
    #  set up initial directory contents
    os.makedirs(os.path.join(tmpDir, 'RPMS', 'noarch'))
    os.makedirs(os.path.join(tmpDir, 'SPECS'))
    os.makedirs(os.path.join(tmpDir, 'BUILD'))
    os.makedirs(os.path.join(tmpDir, 'SOURCES'))
    os.makedirs(os.path.join(tmpDir, 'SRPMS'))

    log.msg(format="Created RPM build structure in %(path)r",
            path=tmpDir)
    return tmpDir



def setupBuildFiles(buildDir, config):
    """
    Create files required to build an RPM in the build directory.
    """
    # Create the source tarball in the SOURCES directory.
    tarballName = "%s.tar" % (config['release-name'],)
    tarballPath = os.path.join(buildDir, "SOURCES", tarballName)
    tarballHandle = tarfile.open(tarballPath, "w")

    sourceDirInfo = tarfile.TarInfo(config['release-name'])
    sourceDirInfo.type = tarfile.DIRTYPE
    sourceDirInfo.mode = 0755
    tarballHandle.addfile(sourceDirInfo)

    tapFileBase = os.path.basename(config['tapfile'])

    initFileInfo = tarfile.TarInfo(
            os.path.join(
                config['release-name'],
                '%s.init' % config['rpmfile'],
            )
        )
    initFileInfo.type = tarfile.REGTYPE
    initFileInfo.mode = 0755
    initFileRealData = initFileData % {
            'tap_file': tapFileBase,
            'rpm_file': config['release-name'],
            'twistd_option': config['twistd_option'],
        }
    initFileInfo.size = len(initFileRealData)
    tarballHandle.addfile(initFileInfo, StringIO(initFileRealData))

    tapFileHandle = open(config['tapfile'], 'rb')
    tapFileInfo = tarballHandle.gettarinfo(
            arcname=os.path.join(config['release-name'], tapFileBase),
            fileobj=tapFileHandle,
        )
    tapFileInfo.mode = 0644
    tarballHandle.addfile(tapFileInfo, tapFileHandle)

    tarballHandle.close()

    log.msg(format="Created dummy source tarball %(tarballPath)r",
            tarballPath=tarballPath)

    # Create the spec file in the SPECS directory.
    specName = "%s.spec" % (config['release-name'],)
    specPath = os.path.join(buildDir, "SPECS", specName)
    specHandle = open(specPath, "w")
    specFileRealData = specFileData % {
            'description': config['description'],
            'rpm_file': config['rpmfile'],
            'version': config['set-version'],
            'tarfile_basename': tarballName,
            'tap_file': tapFileBase,
            'date': time.strftime('%a %b %d %Y', time.localtime(time.time())),
            'maintainer': config['maintainer'],
            'long_description': config['long_description'],
        }
    specHandle.write(specFileRealData)
    specHandle.close()

    log.msg(format="Created RPM spec file %(specPath)r",
            specPath=specPath)

    return specPath



def run(options=None):
    #  parse options
    try:
        config = MyOptions()
        config.parseOptions(options)
    except usage.error, ue:
         sys.exit("%s: %s" % (sys.argv[0], ue))

    #  create RPM build environment
    tmpDir = makeBuildDir()
    specPath = setupBuildFiles(tmpDir, config)

    #  build rpm
    job = subprocess.Popen([
            "rpmbuild",
            "-vv",
            "--define", "_topdir %s" % (tmpDir,),
            "-ba", specPath,
        ], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    stdout, _ = job.communicate()

    # If there was a problem, show people what it was.
    if job.returncode != 0:
        print stdout

    #  copy the RPMs to the local directory
    rpmPath = glob.glob(os.path.join(tmpDir, 'RPMS', 'noarch', '*'))[0]
    srpmPath = glob.glob(os.path.join(tmpDir, 'SRPMS', '*'))[0]
    if not config['quiet']:
        print 'Writing "%s"...' % os.path.basename(rpmPath)
    shutil.copy(rpmPath, '.')
    if not config['quiet']:
        print 'Writing "%s"...' % os.path.basename(srpmPath)
    shutil.copy(srpmPath, '.')

    #  remove the build directory
    shutil.rmtree(tmpDir)

    return [os.path.basename(rpmPath), os.path.basename(srpmPath)]
